instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb

```diff
@@ -0,0 +1,277 @@
+              if headers.key?(key)
+                result[value] = headers[key]
+              else
+                # do case-insensitive match:
+                headers.each do |k, v|
+                  if k.upcase == key
+                    result[value] = v
+                    break
+                  end
+                end
+              end
+            end
+          end
+
+          def untraced_request?(env)
+            return true if untraced_endpoints.include?(env['PATH_INFO'])
+            return true if untraced_requests&.call(env)
+
+            false
+          end
+
+          # https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md#name
+          #
+          # recommendation: span.name(s) should be low-cardinality (e.g.,
+          # strip off query param value, keep param name)
+          #
+          # see http://github.com/open-telemetry/opentelemetry-specification/pull/416/files
+          def create_request_span_name(request)
+            # NOTE: dd-trace-rb has implemented 'quantization' (which lowers url cardinality)
+            #       see Datadog::Quantization::HTTP.url
+
+            if (implementation = url_quantization)
+              request_uri_or_path_info = request.env['REQUEST_URI'] || request.path_info
+              implementation.call(request_uri_or_path_info, request.env)
+            else
+              "HTTP #{request.request_method}"
+            end
+          end
+
+          def extract_remote_context(request)
+            OpenTelemetry.propagation.extract(
+              request.env,
+              getter: OpenTelemetry::Common::Propagation.rack_env_getter
+            )
+          end
+
+          def request_span_attributes(env)
+            attributes = {
+              'http.method' => env['REQUEST_METHOD'],
+              'http.host' => env['HTTP_HOST'] || 'unknown',
+              'http.scheme' => env['rack.url_scheme'],
+              'http.target' => env['QUERY_STRING'].empty? ? env['PATH_INFO'] : "#{env['PATH_INFO']}?#{env['QUERY_STRING']}"
+            }
+
+            attributes['http.user_agent'] = env['HTTP_USER_AGENT'] if env['HTTP_USER_AGENT']
+            attributes.merge!(extract_request_headers(env))
+            attributes
+          end
+
+          def build_attribute_name(prefix, suffix)
+            prefix + suffix.to_s.downcase.gsub(/[-\s]/, '_')
+          end
+
+          def record_frontend_span?
+            config[:record_frontend_span] == true
+          end
+
+          def untraced_endpoints
+            config[:untraced_endpoints].compact
+          end
+
+          def untraced_requests
+            config[:untraced_requests]
+          end
+
+          def url_quantization
+            config[:url_quantization]
+          end
+
+          def response_propagators
+            config[:response_propagators]
+          end
+
+          def allowed_request_headers
```

Comment:
@open-telemetry/ruby-contrib-approvers When I started this implementation I used constructor injection, but then soon realized there are some limitations with that approach that supports _injecting_ this middleware _after_ the Rack app is initialized like in this example: 

https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/rack/example/trace_demonstration2.rb#L27

- Manual instantiation required:  The existing implementation requires users to explicitly add Rack tracer middleware to the builder via `use`. Following the same pattern for the `EventHandler` object instance _may_ be initialized _prior_ to the SDK initialization.  
- Stale config variables: The SDK is lazily initialized so the configuration options may not yet be set on the instrumentation instance. That means the `EventHandler` has to defer looking up the configuration options until _after_ the SDK is initialized, however it does not know when that happens because there aren't any lifecycle hooks that would notify this object instance that the config options are available. Using instance variables made it so the EventHandler had a set of default options and did not work correctly. 
- Tracer not upgraded: The object instance had a stale instance of a tracer, which did not make any sense to me as I expected the tracer to be a upgraded once the SDK was initialized so none of the test passed.

The middleware [instrumentation](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/tracer_middleware.rb) _memoizes_ these options as class variables. That caught me by surprise since I thought it made more sense for these options to be configured/memoized as part of the instrumentation `instance`. 

Is there a case where Rack requests will start to be processed before the SDK will be initialized?

Does the existing rack middleware account for the instrumentation being disabled via environment variables or the config attributes?

Reply:
I moved all of the configuration setup into the `install` block.